### ISSUE TITLE
Custom Taxonomy Routes

### ISSUE THREAD

**Issue by warkior** [statamic/cms#2403]
##  Bug Description
We seem to have lost the ability to use custom taxonomy routes.

I have a taxonomy named `Magazine Topic`. It seems to also suffer from issue #2273. 

On Statamic v2 we had defined a custom route for the taxonomy to use /magazine/topic/{slug}. I don't seem to be able to reproduce it on Statamic v3.

**Statamic version:** 3.0.7

**PHP version:** 7.3

**Comment by mikemartin** [comment]
Hey @warkior 
I agree that there should be a way to customize the taxonomy routes but have you tried naming your taxonomy `topic` and adding that taxonomy to your magazine collection settings? Then the taxonomy url `/magazine/topic/{slug}` should resolve.

**Comment by JonKaric** [comment]
+1 for this. Working with taxonomy routes and having them Just Work™ is a little harder in v3. It would be nice to be able to define a custom route and still have access to the data, like on the default routes, available there too.

**Comment by FrankPeters** [comment]
+1 for this as well. We also heavily relied on custom routing functionality.

**Comment by FrankPeters** [comment]
I have come to the conclusion that we cannot implement the workaround as suggested. So we are now forced to migrate the taxonomy to a collection instead, as it allows us to use a custom routing scheme. 

This is really disappointing as this **breaking change** is not even documented on the https://statamic.dev/upgrade-guide#breaking-changes-core page.

**Comment by JonKaric** [comment]
@FrankPeters You can still define custom routes in your `routes.php` file. You can then use `{taxonomy_term}` in your templates to get the term. It's just a bit more limited than the automatic routes that Statamic provides. 

```
Route::statamic('reviews/category/{taxonomy_term}', 'my_view'); 
```

**Comment by FrankPeters** [comment]
@JonKaric Thanks! It seems to have worked. This will save a ton of work. Still have to solve some issues, but they appear minor now compared to converting to collections ;-)

**Comment by damcclean** [comment]
@FrankPeters Oops, pretty sure it wasn't intentionally left off the list of breaking changes. Thanks for pointing that out though. I've PR'd it into the docs statamic/docs#319.

**Comment by JonKaric** [comment]
Re-opening this

**Comment by 1stevengrant** [comment]
Docs suggestion being able to define the route in the YAML file for the taxonomy https://statamic.dev/knowledge-base/taxonomies-by-hand#creating-taxonomies but that doesn't work 


```
title: 'Stevens Tags'
route: 'abc/sg/{slug}'
```

However when I loop through the tags 
```
{{ stevens_tags }}
        {{ url }}
    {{ /stevens_tags }}
```

the output is still `/stevens_collection/stevens-tags/tag`

**Comment by stephenmeehanuk** [comment]
I'm hitting this same problem. Is there a fix on the horizon?


Is the issue related to the need for a taxonomy subfolder in a collection?

If so, would it be possible to replace the subfolder with `taxonomy.index.antlers.html` and a `taxonomy.show.antlers.html` files in the collection folder instead?

**Comment by jasonvarga** [comment]
You're able to have collection-specific taxonomy routes.

https://statamic.dev/taxonomies#templating

e.g. if you create `resources/views/blog/tags/index.antlers.html`, then `/blog/tags` will be available.
and `resources/views/blog/tags/show.antlers.html` will make `/blog/tags/mytag` will be available.

**Comment by 1stevengrant** [comment]
yeah, that works but if you wanted to customise the route further?

I have a site on v2 that has a particular routing structure they don't want to move from. I've been able to circumvent via some gnarly route file magic.

**Comment by jasonvarga** [comment]
Just pointing that out since some comments sound like they didn't know that was possible.

**Comment by 1stevengrant** [comment]
sure - I've not had any issues on new builds but definitely a quirk on a migration

**Comment by stephenmeehanuk** [comment]
> and `resources/views/blog/tags/show.antlers.html` will make `/blog/tags/mytag` will be available.

I guess I was wondering if it's possible to drop the taxonomy name from the URL?

When viewing `/blog/tags/mytag` could `/blog/mytag` be used instead?

**Comment by jasonvarga** [comment]
@stephenmeehanuk where are your blog posts located? /blog/{slug}? wouldn't that conflict with /blog/{tag}?

**Comment by stephenmeehanuk** [comment]
@jasonvarga I guess there would be a conflict... 

I'm curious how this bit works though. 

Using the example from the docs, on `/blog/tags/retrowave` all the 'retrowave' entries would be listed. But the `url` to each entry would be  something like `/blog/retrowave/entry-post-url-here`. 

Based on the need for the taxonomy slug to be present in the URL for the collection term page, I would expect it to be used in the entry url - but it's not?

**Comment by jasonvarga** [comment]
>  But the url to each entry would be something like `/blog/retrowave/entry-post-url-here`.

Where are you getting that from? The URL of the entry is defined in the `route` in the collection. If it's `route: 'blog/{slug}` then the url would be `/blog/entry-post-url-here`. The entry urls aren't affected by taxonomies.

**Comment by stephenmeehanuk** [comment]
> Where are you getting that from? 

Apologies, I had the wrong end of the stick! I get taxonomy URL and collection URL are two different things.

---

Are there (any near future) plans to bring back custom taxonomy routes as mentioned at the top of this post? Or fixing the issue @1stevengrant mentions? Seems like the docs say custom taxonomy routes are possible, but it's not working?

Currently I need to use this structure when using a taxonomy with a collection `collection-name/taxonomy/term`, it'd be nice to be able to use `collection-name/term` instead.

**Comment by jasonvarga** [comment]
I'd like to support it at some point since people clearly want it. I don't have an ETA though.

I've fixed Steven's issue with the docs by removing it. 😄

**Comment by 1stevengrant** [comment]
😂

**Comment by 1stevengrant** [comment]
@stephenmeehanuk here's how I've worked around it 

```php
Route::get('whats-on/festival/{slug}', function ($slug) {

    $festival = Term::findBySlug($slug, 'festival');

    return (new \Statamic\View\View)
        ->layout('layout')
        ->template('events.festival')
        ->with(['title' => $festival->title()]);
});
```

**Comment by jasonvarga** [comment]
Here's a gist with an example of how to implement custom taxonomy routes. It builds on what Steven posted.

https://gist.github.com/jasonvarga/4342f96b77d52fcd27dad22d88253e1f

**Comment by Oleafeon** [comment]
@jasonvarga this works fine but how do i handle it when its a multisite

**Comment by warkior** [comment]
The various comments above have been very helpful. Here's a way we've solved it for now. We've added the following to our `routes/web.php`

```PHP
Route::get('magazine/topics/{slug}', function ($slug) {

    /** @var \Statamic\Taxonomies\LocalizedTerm $term */
    $term = Term::findBySlug($slug, 'magazine_topic');
    if( !$term )
        return response((new \Statamic\View\View)
            ->layout('layout')
            ->template('errors.404'), 404);

    return (new \Statamic\View\View)
        ->layout('layout')
        ->template('magazine_topic.show')
        ->with([
            'title' => $term->title(),
            'slug' => $slug,
        ] + $term->data()->toArray());

});
```

This allows page loads of `/magazine/topics/something` to load the `something` term, show a custom 404 if it's missing, and show the view with all term data if it exists. For clarity our term handle is actually `magazine_topic`.

We are then able to use the following in `views/magazine_topic/show.antlers.html`.

```HTML
{{ collection from="magazine_articles" taxonomy:magazine_topic="{slug}" paginate="true" as="posts" }}
  ...
  {{ posts }}
    <h3> {{ title }} </h3>
    ...
  {{ /posts }}
  ...
{{ /collection }}
```

Hope this is helpful to someone out there. :)

**Comment by iipavlov** [comment]
I have the strange case where on my development and staging environments the  `collection-name/taxonomy/term` path works fine with the custom templating, but on the production environment same path returns 404 page.
The content, views etc between all environments are synced (merged branches).

Where should I look for the issue?

**Comment by jackmcdade** [comment]
Closing in favor of https://github.com/statamic/ideas/issues/839. We're right on the fence between deprecating Taxonomies entirely and going all in on Collections, or doing the big work. We'll see. But this is an old issue and I'm trying to clean everything up. 😅
